### PR TITLE
#5 .fromFile() does not render

### LINKED ISSUE #5: .fromFile() does not render
Hello!

It seems to me that .fromFile() does not render the content. Did some debugging and found out, that the `shtml` gets retrieved but not rendered though:

Bildschirmfoto 2021-06-25 um 11 13 45

May be an issue concerning some ui5 lifecycle vs. async file loading? 

Would be great if you guys could have a look into it. After some research on my own I am really curious whats the problem behind...

Cheers,
Philipp

FAIL  test/ui5-app/basic.test.js
  Markdown
    ✓ should render markdown via content property (2321 ms)
    ✕ should render markdown via fromFile property (1084 ms)
    ✓ should render markdown via binding (494 ms)

  ● Markdown › should render markdown via fromFile property

    TimeoutError: Element h1 (text: "first") not found

      waiting for function failed: timeout 500ms exceeded

Test Suites: 1 failed, 1 total
Tests:       1 failed, 2 passed, 3 total
Snapshots:   0 total
Time:        4.314 s, estimated 5 s

### PR DESCRIPTION
**Pull request by phreis**
Hello! 👋

fixes #5 

.fromFile() did not render the content. After async file fetching I inserted a `oControl.setContent(...)` to get the control properly rerenderd.  

Regards,
Philipp

### CODE REVIEW

**Comment by wridgeu** [comment]
Hey there @phreis appreciate the PR!

I added @vobu to see if he's ok with the small change made (guess if it works it should be fine :P) however the `.vscode` folder and with it the `settings.json` was left within the project on purpose iirc and shouldn't be removed. 🙂 

BR,
Marco

**Comment by phreis** [comment]
Hi @wridgeu !

Thanks for picking it up! Your are right, pls ignore the deletion of `settings.json` ... should not be committed

BR,
Philipp

**Comment by wridgeu** [comment]
Had to adjust the test timeout to 20s (probably something to do with my machine today ...) other than that it worked (couldn't check properly earlier).

rendered output
image
test run
image

so other than the `.vscode` folder: lgtm!

I'll let @vobu approve it fully. =) 

Thanks again!

BR,
Marco

**Comment by phreis** [comment]
Excluded `.vscode` from `.gitignore` (as it was) - so `.vscode/settings.json` shows up again!

BR,
Philipp

File: Markdown.js [hunk 1]
```diff
@@ -30,12 +30,7 @@ sap.ui.define(["sap/ui/core/Control", "./marked/marked.min"], (Control /*, marke
                 } else if (oControl.getFromFile()) {
                     fetch(oControl.getFromFile())
                         .then((r) => r.text())
-                        .then((md) => {
-                            // eslint-disable-next-line no-undef
-                            const sHtml = marked(md)
```
Comment (vobu) [inline Markdown.js:35]: but where does the actual markdown to html conversion happen then if the call to `marked` is removed?

**Comment by vobu** [comment]
thanks for your work so far, you two! just left a review comment.
will also set up gh action as ci for this (and eventually later for cd &rarr; npm registry as well), so PRs can run through tests

File: Markdown.js (hunk 1 above)
Comment (phreis) [reply Markdown.js:35]: Okay let me see if I can explain the thing: 
The html conversion happens in the conditional code block above once `sMarkdown` gets truthy. 

When the property `content` is passed to the control, it gets rendered once and we are good to go.

When the property `fromFile` is passed to the control, the file is fetched (async!) in the first rendering cycle. Once the promise resolves, this first rendering cycle has already finished. So we need an other cycle which is initiated by `oControl.setContent(md)` In this second cylce `sMarkdown` can be derived from `oControl.getContent()` an the conversion finally happens.

Comment (wridgeu) [reply Markdown.js:35]: I got one more question, more for myself than anything. ^^

Is there a way to better analyze (visualize maybe) the different rendering cycles in the browser dev tools or which approach did you go with when looking into this - f.e. use several breakpoints & reloads?

Comment (vobu) [reply Markdown.js:35]: @phreis: ah, now i get it: you're utilizing an additional rendering cycle triggered by a `$control.setProperty()` in order to make the retrieved markdown available to the renderer logic.
Not sure whether i like this performance-wise but i like the code removal :)

**Review by vobu — APPROVED** [review]
LGTM

File: Markdown.js (hunk 1 above)
Comment (phreis) [reply Markdown.js:35]: First I have to admit that I am still learning UI5  🙂
And yes, your are right I figured out using several breakpoints in chrome dev tool eg. at `if (sMarkdown) {` and in the `.then` branches. 

Further more documentation sais, that `setContent()` marks the control as changed which enforces automatic re-rendering. 
Side note: Invalidation ( and re-rendering) can be suppressed by: `oControl.setProperty('content',md,true`
[https://sapui5.hana.ondemand.com/sdk/#/api/sap.ui.base.ManagedObject%23methods/setProperty](url)

**Pull request merged**